model/service.go

```diff
@@ -21,40 +153,183 @@ import (
 	"strings"
 )
 
-// ServiceDiscovery enumerates Istio service instances
+
+// Service describes an Istio service (e.g., catalog.mystore.com:8080)
+type Service struct {
+	// Hostname of the service, e.g. "catalog.mystore.com"
+	// TODO need a FQDN check in rules validation
+	// Hostname uniquely identifies a service and is used as a reference
+	// identifier.  The hostname depends very much on the underlying
+	// platform. For example in Kubernetes, for a service called myservice
+	// running in namespace foobar, its hostname might be of the form
+	// myservice.foobar.svc.cluster.local
+	Hostname string `json:"hostname"`
+
+	// Address specifies the service IPv4 address of the load balancer (a
+	// virtual IP), if available through the underlying platform. For e.g.,
+	// in Kubernetes, this translates into the Service's cluster wide
+	// Virtual IP.
+	Address string `json:"address,omitempty"`
+
+	// Ports is the set of network ports where the service is listening for
+	// connections.  Depending on the platform, the port can be annotated
+	// with additional information such as the type of protocol used by the
+	// port (e.g., in Kubernetes, one can use the port.protocol field to
+	// specify whether the service is listening on TCP/UDP port, and the
+	// port.name field to indicate whether the service is listening on Http
+	// port or not.
+
+	// FIXME are we obtaining port types like HTTP2, GRPC, etc., from port
+	// names? If so, are we expecting users to follow naming conventions ?
+	Ports PortList `json:"ports,omitempty"`
+}
+
+// Port represents a network port
+type Port struct {
+	// Name ascribes a human readable name for the port object. When a
+	// service has multiple ports, the name field is mandatory (FIXME
+	// really? the scenario we are worried about is a kubernetes loophole,
+	// where a looney user can define two ports for a service such that
+	// both ports have the same port number but a different name This is a
+	// corner case and more like an abuse of kubernetes.
```

Comment:
not sure what you are suggesting. it's the convention that names must be unique and non-empty for > 1 ports.